spec/lib/validators/form_template_yaml_validator_spec.rb

```diff
@@ -0,0 +1,174 @@
+            attributes:
+              label: "Full name"
+              placeholder: "eg. John Smith"
+              description: "What is your full name?"
+            validations:
+              required: true
+              minimum: 2
+              maximum: 100
+        YAML
+
+      it "does not add any errors" do
+        validator.validate(form_template)
+        expect(form_template.errors).to be_empty
+      end
+    end
+
+    context "with invalid YAML" do
+      let(:yaml_content) { "invalid_yaml_string" }
+
+      it "adds an error message for invalid YAML" do
+        validator.validate(form_template)
+        expect(form_template.errors[:template]).to include(
+          I18n.t("form_templates.errors.invalid_yaml"),
+        )
+      end
+    end
+  end
+
+  describe "#check_missing_fields" do
+    context "when type field is missing" do
+      let(:yaml_content) { <<~YAML }
+          - id: name
+            attributes:
+              label: "Full name"
+        YAML
+
+      it "adds an error for missing type field" do
+        validator.validate(form_template)
+        expect(form_template.errors[:template]).to include(
+          I18n.t("form_templates.errors.missing_type"),
+        )
+      end
+    end
+
+    context "when id field is missing" do
+      let(:yaml_content) { <<~YAML }
+          - type: input
+            attributes:
+              label: "Full name"
+        YAML
+
+      it "adds an error for missing id field" do
+        validator.validate(form_template)
+        expect(form_template.errors[:template]).to include(
+          I18n.t("form_templates.errors.missing_id"),
+        )
+      end
+    end
+  end
+
+  describe "#check_allowed_types" do
+    context "when YAML has invalid field types" do
+      let(:yaml_content) { <<~YAML }
+          - type: invalid_type
+            id: name
+            attributes:
+              label: "Full name"
+        YAML
+
+      it "adds an error for invalid field types" do
+        validator.validate(form_template)
+        expect(form_template.errors[:template]).to include(
+          I18n.t(
+            "form_templates.errors.invalid_type",
+            type: "invalid_type",
+            valid_types: FormTemplateYamlValidator::ALLOWED_TYPES.join(", "),
+          ),
+        )
+      end
+    end
+
+    context "when field type is allowed" do
+      let(:yaml_content) { <<~YAML }
+          - type: input
+            id: name
+        YAML
+
+      it "does not add an error for valid field type" do
+        validator.validate(form_template)
+        expect(form_template.errors[:template]).to be_empty
+      end
+    end
+  end
+
+  describe "#check_descriptions_html" do
+    context "when description field has safe HTML" do
+      let(:yaml_content) { <<~YAML }
+          - type: input
+            id: name
+            attributes:
+              label: "Full name"
+              description: "What is your full name? Details <a href='https://test.com'>here</a>."
```

Comment:
Is valid HTML output in the browser? I guess we might want that, for example with links to more instructions. 

Can a rogue admin/moderator post valid HTML that breaks the layout here? (For example, by adding a `<div>` element with a class name that has complex stying in core.)

Reply:
Yes, valid HTML is output. We're validating the HTML using Loofah, their SafeList is here:

https://github.com/flavorjones/loofah/blob/main/lib/loofah/html5/safelist.rb

It is kind of loose, allowing `width` or `font-size` CSS or an `h1` element, but it doesn't allow changing the `position` or complex CSS for example. I'm open to suggestions here, we can have a reduced `SafeList` with just what we want to allow (a link, essentially) or I'd be happy to use another approach entirely too, maybe there's already something in the codebase that I could reuse

Reply:
@pmusaraj Thank you for the review! I've added a couple of tests covering the presence of a label and description.

About the HTML description, we also have the `sanitize` gem available, and we could do something really strict like:

```ruby
    sanitized_html =
      Sanitize.fragment(description, elements: ["a"], attributes: { "a" => %w[href target] })

    is_safe_html = sanitized_html == Loofah.html5_fragment(description).to_s
```

This comparison is not ideal, but I'm unsure how to avoid it. At the same time, I wanted to avoid just saving the sanitized HTML and failing silently without a warning to the admin user - hence this validation.

Reply:
Yeah, I like the sanitized string option, allowing only `a` elements with a `href` and/or `target` makes a lot of sense. I think it's best to start strict and safe here. 

With the correct documentation (i.e. "you can use plain text and links for the description fields...") I'm ok with no errors being output to the user. 

Thanks for the detailed explanation!